### PR TITLE
WebHost: Massive overhaul of options pages

### PR DESCRIPTION
**Pull request by LegendaryLinux**
## What is this fixing or adding?
This is a big one.

I have re-written player-options and weighted-options in Jinja, rather than generating it all in JavaScript. The pages are completely usable without JS now, though JS will provide some nice features.

### What has been done:
- player-options and weighted-options are now created entirely in Jinja
- weighted-options is now a per-game page because 87MB of HTML is ridiculous and my 16-core system couldn't render the page in under fifteen seconds
- Interactive elements are hidden by a <noscript> tag if JS is disabled. This does not prevent the user from accessing any options
  - The randomize buttons will not disable their respective inputs, but the toggle will still work without JS
  - The presets select is hidden
- All option types have been implemented
- Option groups have been implemented, and their divs are collapsible (no js required)
- Implement server-side generation of the YAML export
- Implement "generate single-player game"
- Load the user's previous settings from localStorage, and save settings when the form is submitted
- Implement option presets functionality on player-options
- Allow option groups to have a specified display order (implemented by @alwaysintreble)
- Changed the "Generate Game" button to read "Generate Single-Player Game" to reduce confusion experienced by new users

### Out of scope, but I did it anyway
- Added support for option visibility in simple and complex pages, allowing devs to choose if options should appear nowhere, in player-options only, or on all options pages
- Updated supported-games page to include links to the "Advanced Options" pages
- Updated supported-games page to use details/summary tags for collapsible content
- weighted-options pages now share a theme with their other game pages

### Important changes for world devs:
- The tooltip text rendering has been changed. It is no longer passed through JSON, and so takes its formatting more from the __doc__ comment string now. You should look at your world pages and update the whitespace in your docs to make sure they look how you want.

## How was this tested?
I tested this locally, and generated many, _many_ YAML files.

## If this makes graphical changes, please attach screenshots.
image
image
image
image
image
image

### CODE REVIEW

**Comment by LegendaryLinux** [comment]
I have un-broken weigeted-options, thus making this PR ready for actual review and inclusion.

Note not all option types are included in weighted-options yet because that is technically supposed to be covered in a different PR. One for which I will need to rebuild the weighted-options page. As this PR already includes a ton of changes, I figure it's best to not let scope creep get any worse here.

The next two PRs will be for:
1. This PR
2. Rebuild weighted-settings in Jinja and support all option types properly
3. Allow developers to specify which pages to display options on

Weighted Options now shows the group name above the groups, as seen here:
image

**Comment by LegendaryLinux** [comment]
And we're back to a draft because weighted-options is definitely still very broken.

**Comment by LegendaryLinux** [comment]
Screw it. I'm including the weighted-options re-write in this PR as well. The recent JS re-write has caused me to be completely unfamiliar with the code, so I'm just going to nuke now it rather than learn it just so I can nuke it later.

**Comment by LegendaryLinux** [comment]
I have updated the original post's content to reflect the current state of this PR. Please give that a re-read.

File: WebHostLib/static/assets/playerOptions.js [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+      });
+      return;
+    }
+
+    // Handle normal (text, number, select, etc.) and custom inputs (custom inputs exist with TextChoice only)
+    const normalInput = document.getElementById(optionName);
+    const customInput = document.getElementById(`${optionName}-custom`);
+    const rangeValue = document.getElementById(`${optionName}-value`);
+    const randomizeInput = document.getElementById(`random-${optionName}`);
+    if (optionValue === 'random') {
```
Comment (ScipioWright) [inline WebHostLib/static/assets/playerOptions.js:274]:
```suggestion
    if (optionValue == 'random') {
```

File: worlds/AutoWorld.py [hunk 2]
```diff
@@ -3,16 +3,18 @@
...
-from Options import PerGameCommonOptions
-from BaseClasses import CollectionState
+from Options import ExcludeLocations, ItemLinks, LocalItems, NonLocalItems, PerGameCommonOptions, \
+    PriorityLocations, \
+    StartHints, \
+    StartInventory, StartInventoryPool, StartLocationHints
+from BaseClasses import CollectionState, OptionGroup
 
 if TYPE_CHECKING:
     import random
```
Comment (ScipioWright) [inline worlds/AutoWorld.py:20]: Something's up with random in the tests, and I'm gonna assume it's here, but I'm not sure

File: worlds/AutoWorld.py [hunk 3]
```diff
@@ -3,16 +3,18 @@
 import hashlib
 import logging
 import pathlib
-import random
```
Comment (beauxq) [inline worlds/AutoWorld.py:6]: line 331 uses this

I'm not sure why it's in the `if TYPE_CHECKING` block.

Comment (beauxq) [reply worlds/AutoWorld.py:6]: Out of scope for this PR, but it would be better if it's `from random import Random` to discourage any use of the global `random` state.

File: WebHostLib/misc.py [hunk 4]
```diff
@@ -37,23 +40,25 @@ def start_playing():
...
+def option_presets(game):
+    world = AutoWorldRegister.world_types[game]
+    presets = {}
+
+    if world.web.options_presets:
+        presets = presets | world.web.options_presets
+
+    class SetEncoder(json.JSONEncoder):
+        def default(self, obj):
+            if isinstance(obj, set):
```
Comment (beauxq) [inline WebHostLib/misc.py:54]: For some parts of `Options` we use `frozenset` as a default (because we want immutable defaults).
But this doesn't handle `frozenset`
```python
>>> fs = frozenset()
>>> json.dumps(fs)
Traceback (most recent call last):
...
TypeError: Object of type frozenset is not JSON serializable
>>> isinstance(fs, set)
False
>>> from collections.abc import Set             
>>> isinstance(fs, Set)
True
>>> isinstance(set(), Set)
True
```

```suggestion
            from collections.abc import Set
            if isinstance(obj, Set):
```
`collections.abc.Set` will catch both `set` and `frozenset`

**Comment by beauxq** [comment]
I see this:

> The tooltip text rendering has been changed. It is no longer passed through JSON, and so takes its formatting more from the doc comment string now. You should look at your world pages and update the whitespace in your docs to make sure they look how you want.

But the fact that it doesn't even make sure line breaks aren't in the middle of a word seems like a big burden to put on developers.
I think we should be able to apply some basic formatting to handle some of the basic issues generally, and not have to worry about it in each individual doc string.
ap-doc-render

**Comment by Silvris** [comment]
Looks like rendering of option names have also changed. I think at least word wrap should probably be applied.
image

**Comment by Silvris** [comment]
Seems like the tooltip can get misaligned from the indicator as well.
image

File: WebHostLib/templates/playerOptions/playerOptions.html [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+        <div id="player-options-header">
+            <h1>{{ world_name }}</h1>
+            <h1>Player Options</h1>
+        </div>
+        <p>Choose the options you would like to play with! You may generate a single-player game from this page,
+            or download an options file you can use to participate in a MultiWorld.</p>
+
+        <p>
+            A more advanced options configuration for all games can be found on the
+            <a href="/weighted-options">Weighted options</a> page.
```
Comment (Silvris) [inline WebHostLib/templates/playerOptions/playerOptions.html:40]:
```suggestion
            A more advanced options configuration for all games can be found on the
            <a href="/weighted-options">Weighted options</a> page.
```
This is outdated with the most recent changes. Currently the link just takes you back to the games page, but I think we should probably just reword it to link to the world's advanced options page.

File: WebHostLib/static/styles/playerOptions/playerOptions.css [hunk 6]
```diff
@@ -0,0 +1,319 @@
...
+}
+#player-options table .randomize-button[data-tooltip]::after {
+  left: unset;
+  right: 0;
+}
+#player-options table label {
+  display: block;
+  min-width: 200px;
+  margin-right: 4px;
+  cursor: default;
```
Comment (Silvris) [inline WebHostLib/static/styles/playerOptions/playerOptions.css:282]:
```suggestion
  cursor: default;
  word-break: normal;
```
This should fix the wrapping on the option names and tooltips. Interestingly, this also fixes the misaligned tooltips.

File: WebHostLib/templates/playerOptions/macros.html [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+                {% endfor %}
+            </div>
+        </td>
+    </tr>
+{% endmacro %}
+
+{% macro OptionTitle(option_name, option) %}
+    <label for="{{ option_name }}">
+        {{ option.display_name|default(option_name) }}:
+        <span class="interactive" data-tooltip="{{ (option.__doc__ | default("Please document me!"))|escape }}">(?)</span>
```
Comment (Silvris) [inline WebHostLib/templates/playerOptions/macros.html:253]:
```suggestion
        <span class="interactive" data-tooltip="{{ (option.__doc__.strip().replace('    ', '') | default("Please document me!"))|escape }}">(?)</span>
```
Optional, I believe this should replicate the tooltip formatting of current. Not really a fan of the `replace` though. Probably needs an equivalent one for weighted options.

Comment (beauxq) [reply WebHostLib/templates/playerOptions/macros.html:253]: I think `dedent` is what we want.

https://docs.python.org/3/library/textwrap.html#textwrap.dedent

File: worlds/AutoWorld.py (hunk 3 above)
Comment (LegendaryLinux) [reply worlds/AutoWorld.py:6]: I brought in `from random import Random` and used it in the few places it was directly referenced in AutoWorld.

File: WebHostLib/misc.py (hunk 4 above)
Comment (LegendaryLinux) [reply WebHostLib/misc.py]: I implemented your suggested change to `collections.abs.Set`.

File: WebHostLib/templates/playerOptions/playerOptions.html (hunk 5 above)
Comment (LegendaryLinux) [reply WebHostLib/templates/playerOptions/playerOptions.html]: I fixed the link to point properly at the world's weighted options page.

File: WebHostLib/static/styles/playerOptions/playerOptions.css (hunk 6 above)
Comment (LegendaryLinux) [reply WebHostLib/static/styles/playerOptions/playerOptions.css:282]: Word wrap is working in player-options again. I didn't see an issue in weighted-options, but if you see one, I can adjust it.

Comment (LegendaryLinux) [reply WebHostLib/static/styles/playerOptions/playerOptions.css:282]: Also fixed word-wrap for all tooltips on the site, as it was a top-level problem.

File: WebHostLib/options.py [hunk 8]
```diff
@@ -17,189 +16,190 @@ def create():
...
+        response_data = r.json()
+        return redirect(response_data["url"])
+    else:
+        return r.text
 
-    for game_name, world in AutoWorldRegister.world_types.items():
 
-        all_options: typing.Dict[str, Options.AssembleOptions] = world.options_dataclass.type_hints
+def send_yaml(player_name: str, formatted_options: dict):
+    response = Response(yaml.dump(formatted_options))
```
Comment (remyjette) [inline WebHostLib/options.py:77]: This now sorts the keys of the YAML, as the default for `yaml.dump` is `sort_keys=True`

Personally I would prefer if the YAML maintains the ordering provided by the world and presented on the page. Worlds generally group related options together, so maintaining that in the YAML would improve the experience of manually editing it for players and reviewing it for hosts and async helpers. 

```suggestion
    response = Response(yaml.dump(formatted_options, sort_keys=False))
```

**Comment by remyjette** [comment]
I wonder, should the "item & location options" be visible by default on the standard options page? New players generally won't need them, and they make the page a bit overwhelming with the number of available options there. May be worth considering collapsing the section by default.

File: WebHostLib/templates/playerOptions/macros.html (hunk 7 above)
Comment (LegendaryLinux) [reply WebHostLib/templates/playerOptions/macros.html]: Added the `dedent` filter. Weighted options actually doesn't use tooltips, so it isn't needed there.

File: WebHostLib/options.py (hunk 8 above)
Comment (LegendaryLinux) [reply WebHostLib/options.py]: That's extremely helpful. I was wondering if there was a way to prevent the manual re-ordering. Change pushed.

File: worlds/AutoWorld.py [hunk 9]
```diff
@@ -3,19 +3,21 @@
...
```
[the remark this answers is not in the document]
Comment (LegendaryLinux) [reply worlds/AutoWorld.py:6]: Done.

**Comment by LegendaryLinux** [comment]
I like the idea of keeping things simple for new players. I've updated the pages to expand only the first option group on load.

File: Options.py [hunk 10]
```diff
@@ -1198,15 +1198,20 @@ def dictify_range(option: Range):
...
-            }
+
+            option_groups = {option: option_group.name
+                             for option_group in world.web.option_groups
+                             for option in option_group.options}
+            ordered_groups = ["Game Options"]
+            [ordered_groups.append(group) for group in option_groups.values() if group not in ordered_groups]
+            grouped_options = {group: {} for group in ordered_groups}
+            for option_name, option in world.options_dataclass.type_hints.items():
+                grouped_options[option_groups.get(option, "Game Options")][option_name] = option
```
Comment (Silvris) [inline Options.py:1209]: This appears to remove the ability to hide options from the template yaml, so games relying on that (LttP) currently break.

Comment (LegendaryLinux) [reply Options.py]: Good find. This has been corrected.

File: docs/options api.md [hunk 11]
```diff
@@ -85,6 +85,20 @@ class ExampleWorld(World):
...
+Options may be categorized into groups for display on the WebHost. The name of a group may be defined by adding a
+`group_name` to your option. Option groups are displayed alphabetically on the player-options and weighted-options
+pages. Options without a group name are categorized into a generic "Game Options" group.
+
+```python
+from Options import Toggle
+
+class StartingSword(Toggle):
+    """Adds a sword to your starting inventory"""
+    display_name = 'Start With Sword'
```
Comment (Exempt-Medic) [inline docs/options api.md:98]: These should be double-quotes

**Comment by Exempt-Medic** [comment]
A few pages still link to `weighted-settings`, in particular:
https://archipelago.gg/tutorial/A%20Link%20to%20the%20Past/multiworld/es
https://archipelago.gg/tutorial/A%20Link%20to%20the%20Past/multiworld/fr
https://archipelago.gg/tutorial/Archipelago/advanced_settings/en

File: docs/options api.md (hunk 11 above)
Comment (LegendaryLinux) [reply docs/options api.md]: Ugh. Fine.

**Comment by LegendaryLinux** [comment]
Fixed.

File: WebHostLib/templates/playerOptions/macros.html [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+                {% endfor %}
+            </div>
+        </td>
+    </tr>
+{% endmacro %}
+
+{% macro OptionTitle(option_name, option) %}
+    <label for="{{ option_name }}">
+        {{ option.display_name|default(option_name) }}:
+        <span class="interactive" data-tooltip="{% filter dedent %}{{(option.__doc__ | default("Please document me!"))|escape }}{% endfilter %}">(?)</span>
```
Comment (beauxq) [inline WebHostLib/templates/playerOptions/macros.html:253]: One common way to write doc strings is with the triple quotes on their own line.
```python
"""
documentation text here
"""
```
That newline at the beginning is showing up in a bunch of the tooltips.

```suggestion
        <span class="interactive" data-tooltip="{% filter dedent %}{{(option.__doc__.strip('\n') | default("Please document me!"))|escape }}{% endfilter %}">(?)</span>
```

We don't want to strip all whitespace with `.strip()` because that would mess up the `dedent`

This just removes newlines at the beginning and the end.

Comment (beauxq) [reply WebHostLib/templates/playerOptions/macros.html:253]: https://github.com/ArchipelagoMW/Archipelago/pull/3262

File: docs/options api.md [hunk 13]
```diff
@@ -85,6 +85,20 @@ class ExampleWorld(World):
     options: ExampleGameOptions
 ```
 
+### Option Groups
+Options may be categorized into groups for display on the WebHost. The name of a group may be defined by adding a
+`group_name` to your option. Option groups are displayed alphabetically on the player-options and weighted-options
+pages. Options without a group name are categorized into a generic "Game Options" group.
```
Comment (beauxq) [inline docs/options api.md:91]: Is this outdated information?
I tried doing this and it didn't seem to do anything.

I instead defined the list on my `WebWorld` and it did work.

(I like the `WebWorld` API better anyway.)

Comment (LegendaryLinux) [reply docs/options api.md:91]: Oh yeah, that was the old bad implementation. I thought @alwaysintreble changed the documentation as well when implementing the new, better version. I guess it slipped through the cracks.

File: WebHostLib/templates/playerOptions/macros.html (hunk 12 above)
Comment (LegendaryLinux) [reply WebHostLib/templates/playerOptions/macros.html:253]: Stripping the text before the dedent might mess up the filter, but stripping it after the dedent lets us do the necessary touch-ups.
```py
@app.template_filter("dedent")
def filter_dedent(text: str):
    return dedent(text).strip("\n ")
```

Comment (beauxq) [reply WebHostLib/templates/playerOptions/macros.html:253]: That doesn't solve the problem with the 2nd style of doc string addressed in this: https://github.com/ArchipelagoMW/Archipelago/pull/3262

Comment (beauxq) [reply WebHostLib/templates/playerOptions/macros.html:253]: That solution also supports someone who does want an indented first line, with something like this:
```python
class MyOption(Option):
    """
       My neatly formatted paragraph
    is so beautiful.
    """"
```

File: docs/options api.md (hunk 13 above)
Comment (LegendaryLinux) [reply docs/options api.md]: I have updated the documentation.

Comment (LegendaryLinux) [reply docs/options api.md]: I have updated the documentation.

**Comment by Exempt-Medic** [comment]
References to weighted options/settings that should be fixed:
* [Site Map](https://github.com/ArchipelagoMW/Archipelago/blob/option-groups/WebHostLib/templates/siteMap.html#L27)
* [Options API](https://github.com/ArchipelagoMW/Archipelago/blob/option-groups/docs/options%20api.md#option-groups)
* [Overcooked 2 Setup Guide](https://github.com/ArchipelagoMW/Archipelago/blob/option-groups/worlds/overcooked2/docs/setup_en.md#generate-a-multiworld-game) (This line can be entirely removed because the option now appears on the regular options page)
* [StarCraft 2 Setup Guide](https://github.com/ArchipelagoMW/Archipelago/blob/option-groups/worlds/sc2/docs/setup_en.md#where-do-i-get-a-config-file-aka-yaml-for-this-game) (Again, can be entirely removed)
* [Advanced YAML Guide](https://github.com/ArchipelagoMW/Archipelago/blob/option-groups/worlds/generic/docs/advanced_settings_en.md#advanced-yaml-guide) (Could probably just remove the full sentence mentioning "weighted settings"
* [World API](https://github.com/ArchipelagoMW/Archipelago/blob/option-groups/docs/world%20api.md#documenting-locations) and [World API](https://github.com/ArchipelagoMW/Archipelago/blob/option-groups/docs/world%20api.md#documenting-items)

This is also seems to make both `item_descriptions` and `location_descriptions` not display anywhere, or I at least can't find them on any part of the site. If those are being removed, some documentation needs to be updated

**Comment by Exempt-Medic** [comment]
In the various options where you select items, the item groups a game defined are separated out and put at the top. This does not happen for location groups.

**Comment by Exempt-Medic** [comment]
I'm unsure if this is possible, but it would make sense to have the weighted-options pages for games be able to use presets, or at least, reset to the default options.

**Comment by Exempt-Medic** [comment]
Option Lists should be sorted, probably alphabetically. Navigating long lists with seemingly-random sorts is not viable

**Comment by Exempt-Medic** [comment]
LONG
Really long "words" don't display well

**Comment by Exempt-Medic** [comment]
Screenshot 2024-05-06 090653
The spacing/formatting here could be improved

**Comment by Exempt-Medic** [comment]
Is changing to a preset supposed to reload the page now? It doesn't do that currently, but it does with these changes.

**Comment by Exempt-Medic** [comment]
Final Fantasy on `main` does not appear on the weighted-options page. I assume this is intended, given the way making an FF1 YAML works. With these changes, FF1 has an advanced-options page even though it probably should not.

**Comment by Exempt-Medic** [comment]
I'm not sure what determines the order of games in the `/games` and `/weighted-options` pages, but they don't match and neither one is alphabetical.

File: WebHostLib/templates/weightedOptions/weightedOptions.html [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+            <div class="js-warning-banner">
+                This page has reduced functionality without JavaScript.
+            </div>
+        </noscript>
+
+        <div id="user-message"></div>
+
+        <div id="weighted-options-header">
+            <h1>{{ world_name }}</h1>
+            <h1>Weighted Options</h1>
```
Comment (alwaysintreble) [inline WebHostLib/templates/weightedOptions/weightedOptions.html:31]: These titles blend together, and it's really easy to skip over the explanation of the page because of it. Probably a good idea to put another header just above the name section to differentiate it from the explanations.
```suggestion
            <br />
            <h2>Weighted Options</h2>
```

**Review by alwaysintreble — CHANGES_REQUESTED** [review]
Python all seems good, just a few comments on the html. Didn't look at the JS since idk the language nearly well enough to judge it.

File: worlds/AutoWorld.py [hunk 15]
```diff
@@ -118,6 +120,29 @@ def __new__(mcs, name: str, bases: Tuple[type, ...], dct: Dict[str, Any]) -> Aut
...
+        option_groups: List[OptionGroup] = dct.get("option_groups", [])
+        item_and_loc_options = [LocalItems, NonLocalItems, StartInventory, StartInventoryPool, StartHints,
+                                StartLocationHints, ExcludeLocations, PriorityLocations, ItemLinks]
+        seen_options = []
+        item_group_in_list = False
+        for group in option_groups:
+            assert group.name != "Game Options", "Game Options is a pre-determined group and can not be defined."
+            if group.name == "Item & Location Options":
+                group.options.extend(item_and_loc_options)
+                item_group_in_list = True
```
Comment (alwaysintreble) [inline worlds/AutoWorld.py:136]: i didn't notice this before (and it's my bug) but you can put one of the item_and_loc_options in a different group and it won't render correctly. possible solution
```suggestion
            if group.name == "Item & Location Options":
                group.options.extend(item_and_loc_options)
                item_group_in_list = True
            else:
                for option in group.options:
                    assert option not in item_and_loc_options, f"{option} cannot be moved out of the \"Item & Location Options\" Group"
```

File: WebHostLib/templates/weightedOptions/weightedOptions.html (hunk 14 above)
Comment (LegendaryLinux) [reply WebHostLib/templates/weightedOptions/weightedOptions.html:31]: I think I've done all I'm going to stylistically until the PR is accepted. Unless something is broken, the styles will remain as they are.

**Comment by Exempt-Medic** [comment]
Sudoku's option page is now listed as an external link, it is not an external link however

**Comment by ThePhar** [comment]
Hard to check the code specifically at this moment, but in Rogue Legacy options for example, when loading the "Unknown Fate" preset, the NamedRange of Progression Balancing is not setting to 0 correctly as defined in the preset.

Also the text shows as "disabled" near the range slider, but still exports a 50.

image
image

Also appears the preset name no longer outputs in the description YAML.

**Comment by Zunawe** [comment]
For [2e94d7a](https://github.com/ArchipelagoMW/Archipelago/pull/2614/commits/2e94d7a9bb9db9222041d54459af6973defed221), I think world devs should be allowed to specify the order of keys. Alphabetical is a good catch-all if the order is going to be arbitrary anyway, but I think it's reasonable to have orders other than that. For Emerald's `AllowedLegendaryHuntEncounters`, I group them by the method used to access them. Kyogre and Groudon are both found in the same way. And Mew, Deoxys, Ho-Oh, and Lugia are found in the same way. So it's easier to see what you're doing when they're next to each other. I don't have a concrete example, but I can imagine an `OptionSet` for named missions that would be more reasonably ordered from first to last rather than alphabetically.

So my preference is for either the automatic sort to be done only if `valid_keys` is unordered, or to have the webhost continue to take what it's given and give the power and responsibility to worlds, which will probably catch up quickly once people start looking at the new option pages.

File: WebHostLib/misc.py [hunk 16]
```diff
@@ -37,23 +40,26 @@ def start_playing():
...
-    return render_template("weighted-options.html")
-
-
-# Player options pages
-@app.route("/games/<string:game>/player-options")
+@app.route("/games/<string:game>/option-presets", methods=["GET"])
 @cache.cached()
-def player_options(game: str):
-    return render_template("player-options.html", game=game, theme=get_world_theme(game))
+def option_presets(game):
```
Comment (beauxq) [inline WebHostLib/misc.py:45]:
```suggestion
def option_presets(game: str) -> Response:
```

Comment (LegendaryLinux) [reply WebHostLib/misc.py]: Done, and I moved the route into `options.py`, to better categorize it with other relevant routes.

**Comment by LegendaryLinux** [comment]
> Hard to check the code specifically at this moment, but in Rogue Legacy options for example, when loading the "Limited Potential" preset, the NamedRange of Progression Balancing is not setting to 0 correctly as defined in the preset.
> 
> Also the text shows as "disabled" near the range slider, but still exports a 50.
> 
> image image
> 
> Also appears the preset name no longer outputs in the description YAML.

This has been fixed.

File: WebHostLib/templates/playerOptions/macros.html [hunk 17]
```diff
@@ -0,0 +1,264 @@
...
+                {{ RandomizeButton(option_name, option) }}
+                </div>
+                <input type="text" id="{{ option_name }}-custom" name="{{ option_name }}-custom" data-option-name="{{ option_name }}" placeholder="Custom value..." />
+            </div>
+        </td>
+    </tr>
+{% endmacro %}
+
+{% macro PlandoBosses(option_name, option) %}
+    <!-- PlandoBosses is handled by its parent, TextChoice -->
```
[the remark this answers is not in the document]
Comment (LegendaryLinux) [reply WebHostLib/templates/playerOptions/macros.html]: These were leftovers from when I was first trying to figure out how to properly use Jinja. They have been removed.

**Comment by LegendaryLinux** [comment]
> So my preference is for either the automatic sort to be done only if valid_keys is unordered.
@Zunawe, I'm know just enough Python to be dangerous. How would I accomplish this? My understanding is Lists are always unordered, or rather, they have the order they were defined with.

**Comment by Zunawe** [comment]
Sorting looks good to me now. Thanks.

**Comment by ScipioWright** [comment]
The option selection bars are really wide, and the the space for option names is much smaller now, which I really don't like.
Current: 
image

This PR:
image

**Comment by hatkirby** [comment]
There seems to be something off with the way the option pages handle the width of the browser. I'm seeing a horizontal scrollbar on every game's options page, and there's nothing to the right (presumably it is just extra space that would be used to center the page).

 image

**Comment by LegendaryLinux** [comment]
@ScipioWright I bumped the width of the option name column up a bit.
@hatkirby, Is that Firefox? I don't see that width issue in Chrome or Firefox.

**Comment by hatkirby** [comment]
Firefox 126.0 on Windows, at 1210px width.

**Comment by LegendaryLinux** [comment]
I found and fixed the issue by adding `overflow-x: hidden` to the html style for that page. It aligns properly and has no content over there. I have no clue what was causing it. The issue was not present on the weighted-options or landing pages.

**Comment by Ishigh1** [comment]
I tested this branch on my Stardew Valley Tilesanity branch that adds 30000 unique items and locations. With it, the option page just crashes after trying to load the options for every item and locations and taking all of RAM.
It is an extreme example and I don't intend to merge my branch into core (for a while?) anyway, but I think it's good to keep in mind that this limits the amount of locations and items a world can have before its option page breaks
Well, maybe things could be solved by only loading the item/location option part on demand, if there is few enough items and locations, or maybe we'll never accept a world with that many items and locations into core, I don't know, but I thought this should be considered at least
